### PR TITLE
Scribes on k-means

### PR DESCRIPTION
**Pull request by huberf**
The PR should contain all needed additions for the Scribe task, but we did have a question which we wanted your input on before we committed:
As the k-means topic coverage spans multiple sections, we kept our additions in two separate files. Which sections (if any) do you all suggest we put our additions into?
In line with this, is it right that our additions should be marked by a `\subsection` rather than a `\section`?

Thanks!

### CODE REVIEW

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 1]
```diff
@@ -0,0 +1,207 @@
+% Contributers: Noah Huber-Feely, Samuel Sharpe
+\subsection{(1+$\epsilon$)-approximate $k$-means clustering}
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: This should be \section{}

File: chapter_1/chapter_1.tex [hunk 2]
```diff
@@ -6,6 +6,8 @@ \chapter{Centroid-Based Clustering}
   \input{chapter_1/4_kmeans}
   \input{chapter_1/5_hardness}
   \input{chapter_1/6_kmeans_pp}
+  \input{chapter_1/kmeans_1pluseps_approx}
+  \input{chapter_1/kmeans_hardness_d2_k2}
```
Comment (all2187) [inline chapter_1/chapter_1.tex]: Remove this and move to the section 5 file.

File: chapter_1/chapter_1.tex [hunk 3]
```diff
@@ -6,6 +6,8 @@ \chapter{Centroid-Based Clustering}
   \input{chapter_1/4_kmeans}
   \input{chapter_1/5_hardness}
   \input{chapter_1/6_kmeans_pp}
+  \input{chapter_1/kmeans_1pluseps_approx}
```
Comment (all2187) [inline chapter_1/chapter_1.tex]: Prepend with 7 or 6 or whatever section number it ends up being, edit other section numbers as needed.

File: chapter_1/chapter_1.tex [hunk 4]
```diff
@@ -6,6 +6,8 @@ \chapter{Centroid-Based Clustering}
...
+  \input{chapter_1/kmeans_hardness_d2_k2}
   \input{chapter_1/further_reading}
```
Comment (all2187) [inline chapter_1/chapter_1.tex:10]: Could you prepend the section number here too while you are at it (and modify the file name as needed).

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 5]
```diff
@@ -0,0 +1,150 @@
+% Contributers: Carolina Zheng
+\section{Hardness of $k$-means when $k=2$}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Make it subsection

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+So Equation 1 corresponds to the form given in the paper, i.e.
+\begin{align}
+    \text{cost} &= \sum_{e\in E(P,Q)}[(p-1)\frac{1}{p^2}+\left(1-\frac{1}{p}\right)^2+(q-1)\frac{1}{q^2}+\left(1-\frac{1}{q}\right)^2] + \sum_{e\not\in E(P,Q)}2 \\
+    &= \left(2-\frac{1}{p}-\frac{1}{q}\right)|E(P,Q)|+2|E(P,P)|+2|E(Q,Q)| \\
+    &= 2|E| - \frac{n}{pq}|E(P,Q)|
+\end{align}
+
+Minimizing this cost is equivalent to maximizing the densest cut objective $|E(P,Q)|/(|P||Q|)$.\qed
+
+\section{Hardness of $k$-means when $d=2$}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Make it subsection

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+We will prove that $k$-means is NP-hard when $d=2$ and $k=\Theta(n^\epsilon)$ for $\epsilon>0$. We'll try to provide some intuition around the proof structure as well as a summary of the results presented in (CITE).
+
+\subsection{Setup}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Just make it a nice texbf heading, no need to make it a section/subsection that appears in the ToC

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+$$\text{cost}(C)=\frac{1}{\sum_{x\in C} w(x)}\sum_{\{x,y\}\in\binom{C}{2}}w(x)w(y)\|x-y\|^2$$
+
+The reduction will be from Exact Cover by 3-Sets (X3C), which is known to be NP-complete (CITE).\\
+
+\begin{definition}[Exact cover by 3-sets]
+    Given a finite set $U$ containing exactly $3n$ elements and a collection $\cC=\{S_1,\dots,S_l\}$ of subsets of $U$, each of which contains exactly 3 elements, are there $n$ sets in $\cC$ such that their union is $U$?
+\end{definition}
+
+\subsection{Overview}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Just make it a nice texbf heading, no need to make it a section/subsection that appears in the ToC

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+$k' \overset{\text{\tiny d}}{=} k+n^\epsilon=\Theta(n^\epsilon)$ centers.
+
+Suppose $\gamma>\epsilon$. Add $n^{\gamma/\epsilon}$ points very close to each
+other and far from the original instance, and 1 center. The optimal solution
+will use the new center to cluster the new points, and the optimum in the
+original instance will not change. Therefore this is a hard instance with
+$m \overset{\text{\tiny d}}{=} n+n^{\gamma/\epsilon}=\Theta(n^{\gamma/\epsilon})$ points and
+$k' \overset{\text{\tiny d}}{=} k+1=\Theta(n^\gamma)=\Theta(m^\epsilon)$ centers.\qed
+
+\subsection{Walking through the proof}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Just make it a nice texbf heading, no need to make it a section/subsection that appears in the ToC

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 10]
```diff
@@ -0,0 +1,150 @@
+% Contributers: Carolina Zheng
+\section{Hardness of $k$-means when $k=2$}
+We will prove that $k$-means is NP-hard when $k=2$. There is a simple proof by (CITE) that is a reduction from densest cut. Densest cut is equivalent to sparsest cut on the complement graph, which is known to be NP-complete (CITE).\\
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: You are showing it is NP-hard EVEN when k=2. You've already showed it NP hard in class.

Comment (all2187) [reply chapter_1/kmeans_hardness_d2_k2.tex]: Forgot to add the citations

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+\begin{definition}[Densest cut]
+    For a given graph $G=(V,E)$, find a bipartition $(P,Q)$ of the vertices in $G$ that maximizes $|E(P,Q)/(|P||Q|)$, where $E(P,Q)$ denotes the edge set of the cut.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Missing cardinality bar. Use \abs{} macro instead.

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+We will follow the proof of Theorem 1 in the paper and explain the steps.\\
+
+\begin{theorem}
+    $k$-means clustering in general dimension is NP-hard for $k=2$.
+\end{theorem}
+
+Suppose we wish to compute the densest cut for a graph $G$ with no parallel edges. Define the $|V|\times|E|$ matrix $M$ as follows: entry $M[v,e]$ is zero if $e\in E$ is not incident on $v\in V$; otherwise, it is $+1$ for one endpoint $e$ and $-1$ for the other (it does not matter which endpoint is which).
+
+Run $k$-means with $k=2$ on the set of points corresponding to the rows in $M$, of dimension $|E|$. Note that each point represents a vertex in $G$. Given an optimal solution to this $k$-means instance, we claim that two returned clusters correspond to the optimal bipartition $(P,Q)$ in densest cut. It is sufficient to show that minimizing the $k$-means cost on this instance is equivalent to maximizing the densest cut objective.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Specify that each point has dim |E| (i read it as there are |E| points).

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+Now let's compute the $k$-means cost. Let $c(v)=p$ if $v\in P$ and $q$ if $v\in Q$. We have
+\begin{align}
+    \text{cost}=\sum_{v\in P}\|M[v,:]-\mu^p\|^2+\sum_{v\in Q}\|M[v,:]-\mu^q\|^2 &= \sum_{e=1}^{|E|}\sum_{v=1}^{|V|}(M[v,e]-\mu_e^{c(v)})^2
+\end{align}
+
+Suppose $e\not\in E(P,Q)$. Fix $v$. Then $\mu_e^{c(v)}=0$, and $M[v,e]$ is nonzero iff $v$ is incident on $e$; in that case it is either $1$ or $-1$. Therefore $e$ contributes a total cost of 2.
+
+Suppose $e\in E(P,Q)$. Fix $v$. Suppose $v$ is not incident on $e$. Then $M[v,e]=0$ and $\mu^{c(v)}_e=\frac{1}{c(v)}$, so the cost is $\frac{1}{c(v)^2}$. There are $p-1$ such $v\in P$ and $q-1$ such $v\in Q$, so the total cost is $(p-1)\frac{1}{p^2}+(q-1)\frac{1}{q^2}$. Suppose $v$ is incident on $e$. There are two such $v$, one of which is in $P$ and one of which is in $Q$. If $M[v,e]=1$, then $\mu^{c(v)}_e=\frac{1}{c(v)}$, and if $M[v,e]=-1$, then $\mu^{c(v)}_e=-\frac{1}{c(v)}$. Therefore the total cost is $(1-\frac{1}{p})^2+(1-\frac{1}{q})^2$.
+
+So Equation 1 corresponds to the form given in the paper, i.e.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: What equation 1? Use \label and \eqref if trying to refer to an equation you wrote

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+\begin{align}
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Here i'm a bit confused, is the cost you are mentioning the same cost as above (kmeans cost) or are you computing the densest cut cost?

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    \text{cost} &= \sum_{e\in E(P,Q)}[(p-1)\frac{1}{p^2}+\left(1-\frac{1}{p}\right)^2+(q-1)\frac{1}{q^2}+\left(1-\frac{1}{q}\right)^2] + \sum_{e\not\in E(P,Q)}2 \\
+    &= \left(2-\frac{1}{p}-\frac{1}{q}\right)|E(P,Q)|+2|E(P,P)|+2|E(Q,Q)| \\
+    &= 2|E| - \frac{n}{pq}|E(P,Q)|
+\end{align}
+
+Minimizing this cost is equivalent to maximizing the densest cut objective $|E(P,Q)|/(|P||Q|)$.\qed
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Something here isn't clear, i was following until my last comment, but it is unclear what cost you are computing, what "form given in the paper" you are referring to, or how the computer cost here shows what you wanted to show. More details have to be given. Why was it important that the total cost was (1-1/p)^2+(1-1/q)^2 before?

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+\section{Hardness of $k$-means when $d=2$}
+
+We will prove that $k$-means is NP-hard when $d=2$ and $k=\Theta(n^\epsilon)$ for $\epsilon>0$. We'll try to provide some intuition around the proof structure as well as a summary of the results presented in (CITE).
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Same as before, i think the word "even" is important. Might want to mention why in previous section or this one. Also you left the (CITE)

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+\begin{definition}[Decisional weighted $k$-means]
+    Given a multiset $S\subseteq \R^d$, an integer $k$ and $L\in\R$, is there a subset $T\subset\R^d$ with $|T|=k$ such that $\sum_{x\in S}\min_{t\in T}\|x-t\|^2\le L$?
+\end{definition}
+
+Analogously to what was noted in class, we can restate the cost of a cluster $C$ in terms of the points belonging to $C$, without referring to $T$. Specifically, let $w(x)$ denote the weight of point $x$. Then
+
+$$\text{cost}(C)=\frac{1}{\sum_{x\in C} w(x)}\sum_{\{x,y\}\in\binom{C}{2}}w(x)w(y)\|x-y\|^2$$
+
+The reduction will be from Exact Cover by 3-Sets (X3C), which is known to be NP-complete (CITE).\\
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: CITE

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+\begin{definition}[Exact cover by 3-sets]
+    Given a finite set $U$ containing exactly $3n$ elements and a collection $\cC=\{S_1,\dots,S_l\}$ of subsets of $U$, each of which contains exactly 3 elements, are there $n$ sets in $\cC$ such that their union is $U$?
+\end{definition}
+
+\subsection{Overview}
+
+The bulk of the paper goes to proving the following theorem.\\
+
+\begin{theorem}
+The $k$-means clustering problem is NP-hard for $k=\Theta(n^\gamma)$ for some $0<\gamma<1$.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: For d=2 no?

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+Then we add some more points $X$ to $H_{l,n}$ to create $G_{l,n}\cup X$. These
+new points are related to the elements in $S_i$ for $S_i\in\cC$. We then argue
+that iff the X3C instance is satisfiable, these new points can be clustered so
+that the total cost increases by an amount $L_2+(l-n)\alpha$ ($L_2$ defined in
+terms of $l,n$). Therefore our cost limit in the decision problem will be $L_1+L_2-n\alpha$.
+
+We will explain the above more rigorously in the next section. Given this construction, the $k$-means instance $G_{l,n}$ will contain exactly $l(6n+3)+(l-1)9n$ points, for which $k=\Theta(n^\gamma)$ for some $0<\gamma<1$ (according to the magic quantity for $k$ mentioned above; also note that the second $n$ here refers to the number of points in the $k$-means instance, not the $n$ in the 3XC instance). There is a simple reduction to the more general $k=\Theta(n^\epsilon)$ for $\epsilon>0$.\\
+
+\begin{theorem}
+    Suppose Theorem 1 is true. Then the $k$-means clustering problem is also NP-hard for $k=\Theta(n^\epsilon)$ for any $\epsilon>0$.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Note theorem 1, theorem 1.8.3 or whatever, use ref and label

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+\end{theorem}
+
+We reproduce the proof almost verbatim from the paper. Fix $\epsilon>0$, and take a hard instance with $n$ points and $k$ centers, with $k=\Theta(n^\gamma)$ as in Theorem 1.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: ditto as above

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+$m \overset{\text{\tiny d}}{=} n+n^\epsilon=\Theta(n)$ points and
+$k' \overset{\text{\tiny d}}{=} k+n^\epsilon=\Theta(n^\epsilon)$ centers.
+
+Suppose $\gamma>\epsilon$. Add $n^{\gamma/\epsilon}$ points very close to each
+other and far from the original instance, and 1 center. The optimal solution
+will use the new center to cluster the new points, and the optimum in the
+original instance will not change. Therefore this is a hard instance with
+$m \overset{\text{\tiny d}}{=} n+n^{\gamma/\epsilon}=\Theta(n^{\gamma/\epsilon})$ points and
+$k' \overset{\text{\tiny d}}{=} k+1=\Theta(n^\gamma)=\Theta(m^\epsilon)$ centers.\qed
+
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: What about gamma=epsilon?

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+\subsection{Walking through the proof}
+
+Now we will present the paper's proof of Theorem 1. Instead of re-typing everything, we will instead try to expand on explanations and provide some more intuition, and refer the reader to the paper for the technical details.
+
+We will start by reading the paper's description of the grid $H_{l,n}$ under section 2, Reduction, as well as \textbf{Definitions 4 and 5 and Lemma 6}.
+
+Definition 5 allows us to see where the number of clusters $k=l(3n+2)+(l-1)3n$ comes from. A nice clustering corresponds naturally to this $k$, since there are $l-1$ $M$ rows, each with $3n$ points in singleton clusters, and $l$ $R$ rows, with $3n+2$ clusters each, corresponding to $3n+1$ clusters with two consecutive points and 1 cluster with one point (whether $s_i$ or $f_i$ is the singleton cluster depends on whether $R_i$ is in an $A$ or $B$ clustering, respectively). Note an $A$ clustering saves exactly $\alpha$ cost compared to a $B$ clustering.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Where are def 4 and lemma 6 discussed?

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+Suppose $\gamma>\epsilon$. Add $n^{\gamma/\epsilon}$ points very close to each
+other and far from the original instance, and 1 center. The optimal solution
+will use the new center to cluster the new points, and the optimum in the
+original instance will not change. Therefore this is a hard instance with
+$m \overset{\text{\tiny d}}{=} n+n^{\gamma/\epsilon}=\Theta(n^{\gamma/\epsilon})$ points and
+$k' \overset{\text{\tiny d}}{=} k+1=\Theta(n^\gamma)=\Theta(m^\epsilon)$ centers.\qed
+
+\subsection{Walking through the proof}
+
+Now we will present the paper's proof of Theorem 1. Instead of re-typing everything, we will instead try to expand on explanations and provide some more intuition, and refer the reader to the paper for the technical details.
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: This whole section is a bit hard to read without having read the paper before hand. Giving a summary/intuition and omitting technical details is totally fine (and ideal). It makes sense to refer readers to the paper for technical details. However, try to make the intuition/summary self contained. Here i dont know what definitions 4 and lemma 6 are. Nor do i know precisely what lemma 7 is, etc. The "let's read up until" comments make this more of a reading along guide than a self contained summary. Try to edit it a bit to make it closer to the latter (although again referring readers to the paper for technical details is what we want to do).

File: chapter_1/kmeans_hardness_d2_k2.tex [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+
+Definition 8 and Lemma 9 justify the choice of $L_2=6n(l-1)h^2\frac{2w}{2w+1}$; looking at the figure, we can see that the ``good" clusters mentioned in Lemma 9 are indeed the closest clusters, and minimize the cost of each of the $6n(l-1)$ new points to be $h^2\frac{2w}{2w+1}$, if they are added to one of these existing clusters.
+
+\textbf{Lemma 10} goes further and states that any optimal $k$-clustering of $G_{l,n}\cup X$ must retain the ``nice'' clustering on the $H_{l,n}$ instance, and further, each $x\in X$ must be assigned to a different ``good'' cluster. The proofs of Lemmas 9 and 10 simply add up the costs according to the construction.
+
+Now we're ready for the reduction. I think this part was the trickiest to understand, so I'll try to provide some intuition.
+
+\textbf{Lemma 11.} \textit{The set $G_{l,n}\cup X$ has a $k$-clustering of cost less or equal to $L$ if and only if there is an exact cover $\mathcal{F}\subseteq \cC$ for the X3C instance.}
+
+$(\implies)$ Suppose we have an optimal $k$-clustering with cost $\le L$. The
```
Comment (all2187) [inline chapter_1/kmeans_hardness_d2_k2.tex]: Is this the proof of lemma 11? If so put it in \begin{proof} environment or just state you will now prove lemma 11. It's a little unclear as is.

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+$\sum_{x \in S} ||x-c(S)||^2$. Here $c(s)=(1/|S|) \sum_{x \in S} x$ and is
+thus the centroid of the set $S$. Let diam($X$) denote the diameter of a set
+$X \subset \bbR^d$. Further, this and other centroid base clustering methods relies upon the Vornoi partition of a set of points by the centroid points.
+For centroids $c_1,...,c_k$, we refer to the Vornoi partition $(S_1,S_2,...,S_k)$ of a set
+as $\Pi_{Vor}(c_1, c_2, \dots, c_k))$.
+
+We say that a $k$-clustering $\Pi$ of $X$ is (1 + $\epsilon$)-approximately optimal if cost($\Pi$) $\leq$ (1 + $\epsilon$)cost($\Pi'$) for any $k$-clustering $\Pi'$ of $X$. The method described below thus achieves the results outlined in the following two theorems.
+
+\begin{theorem}
+Let $X \subset \bbR^d$ be an n-point set, and let $\epsilon >0$ be given. A (1 + $\epsilon$)-approximately optimal 2-clustering of X can be found in time $O(n\text{ log }n \cdot \epsilon^{-2d} \text{log }\frac{1}{\epsilon} + n \epsilon^{-(4d-2))}\text{log }\frac{1}{\epsilon})$ so the running time is $O(n\text{ log }n)$ for a fixed $\epsilon>0$.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Use \R instead of \bbR

Comment (all2187) [reply chapter_1/kmeans_1pluseps_approx.tex]: Use \log

Comment (all2187) [reply chapter_1/kmeans_1pluseps_approx.tex]: Do this throughout (many occurences below)

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+
+\begin{theorem}
+Let $X \subset \bbR^d$ be an n-point set, let $k \geq 3$ be fixed and let $\epsilon >0$ be given. A (1 + $\epsilon$)-approximately optimal k-clustering of X can be found in time $O(n(\text{log }n)^k \epsilon^{-2k^2d})$.
+\end{theorem}
+
+We thus see that we achieve a near-linear time algorithm for both (1 + $\epsilon$)-approximate 2-means and k-means.
+
+This approach requires some additional preliminary terminology and cost definitions which are themselves interesting ideas for this problem space.
+\begin{definition}[$\epsilon$-near]
+Let $\epsilon \geq 0$. We define $\sim_{\epsilon}$ to be a relation on ordered pairs of points in $\bbR^d$. We say that $(x,y)$ and $(x',y')$ are $\epsilon$-near if $(x,y) \sim_{\epsilon} (x', y')$ such that  $(x,y) \sim_{\epsilon} (x', y')$ if $||x-x'|| \leq \epsilon \cdot ||x-y||$ and $||y - y'|| \leq \epsilon \cdot ||x - y||$.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Use \norm{} macro throughout

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+\end{definition}
+
+\begin{definition}[$\epsilon$-separated]
+Let $P$ be a set of ordered pairs of points of $\bbR^d$. We say $P$ is $\epsilon$-separated if no two pairs in $P$ are $\epsilon$-near.
+\end{definition}
+
+\noindent \emph{Approximate range searching}. An important and useful concept for optimizing this problem is approximate range searching. At a high level, this method uses intelligent preprocessing to attempt to find an approximate answer to the weight sum within a specific range for a set $P \subset \bbR^d$ where each point is assigned a weight such that for $p \in P$, $w(p) \in S$ where $S$ is a commmutative semigroup (meaning weights can be summed). Formally, given a range $R \in \cR$ where $\cR$ is a class of admissible ranges and some $\epsilon>0$, we define $R^+$ to be the set of all points at most a distance of $\epsilon \cdot$diam($R$) from $R$, and let $R^-$ be all points of distance at most $\epsilon \cdot$diam($R$) from the complement of $R$.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: weight**ed** sum?

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 28]
```diff
@@ -0,0 +1,207 @@
...
+
+\begin{definition}[$\epsilon$-approximate answer]
+An $\epsilon$-approximate answer to the query with range $R$ is any weight of the form $\sum_{p \in P_1} w(p)$, where $P_1$ is any $\epsilon$-aproximate intersection of $P$ with $R$.
+\end{definition}
+
+This method takes $O(n\text{ log }n)$ preprocessing time and can then produce an $\epsilon$-approximate answer in $O(\text{log}n + \epsilon^{-d})$ as shown by Arya and Mount.
+
+\noindent \emph{Putting the points on a polynomial-size grid}. The proposed k-clustering problem works with points with polynomially large distances between them, however ratios of interpoint distances can be arbitrarily large. The set of points $X$ can be 'snapped' onto a integer grid with polynomial size (specifically $O(n^3/\epsilon)$). If there is some algorithm that can compute an $(1+\epsilon)$-approximate clustering for $X'$. Then a $(1+\epsilon)$-approximate clustering for the arbitrarily large $X$ can be computed with $O(nlogn)$ processing time. This is tangential to the main result so we do not present the proof here. Now we proceed assuming the n-point set $X \subset \bbR^d$ lie on an integer grid of size $O(n^3/\epsilon)$.\\
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: For quotaions in latex use ` and '

Comment (all2187) [reply chapter_1/kmeans_1pluseps_approx.tex]: \log not log

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 29]
```diff
@@ -0,0 +1,207 @@
...
+
+\noindent \textbf{Approximate Centroid Sets}\\
+Let $S$ be a finite set in $\bbR^d$. The quadratic-mean radius of S is
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: $S$ not S

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 30]
```diff
@@ -0,0 +1,207 @@
...
+\begin{align}
+    \rho(S)  = \left( \frac{1}{|S|}\sum_{x\in S}\norm{x-c(S)}^2\right)^{1/2}
+    = \sqrt{\frac{cost(S)}{\abs{S}}}
+\end{align} 
+be the \emph{quadratic-mean radius} of S. For a real number $\epsilon \geq 0$, the \emph{$\epsilon$-tolerance ball} of S is the ball centered as $c(S)$ of radius $(\epsilon/3)\rho(S)$.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Ditto as previous

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 31]
```diff
@@ -0,0 +1,207 @@
...
+ 
+To construct this set we will reference \emph{cubes} which are axis-parallel cubes in $\bbR^d$ and the $K$-enlargement of a cube is a concentric cube with a side $K$-times larger. We let $Q_0$ be the 3-enlargement of the smallest cube enclosing $X$, and $R$ be the side of $Q_0$. We define an \emph{active cube} as any cube containing at least $x/2^{d+1}$ points of X. We initially set $C = \emptyset$. 
+
+\noindent The construction of C begins at each step by repeated subdivision of $Q$ into $2^d$ equal-sized axis aligned cubes. For any cube $Q$ that is active, we let $\sigma$ be the side length of $Q$ and choose $C_Q$ that is $(\frac{\epsilon }{18} \sigma)$-dense for the 2-enlargement of $Q$ (see Figure \ref{fig:cq}). We can say $|C_Q| = O(\epsilon^{-d})$. Finally, we add $C_Q$, to the constructed set C.  After this is done for $Q$, it is no longer active and is further subdivided as long as $\sigma \geq 2r$. We stop when there are no active cubes left. 
+
+\begin{figure}
+    \centering
+    \includegraphics[width=10cm]{chapter_1/files/Cq.png}
+\centering
+    \caption{(a) The set $C_Q$ and (b) the subdivision of $Q$}
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: You should cite the source of this figure

Comment (all2187) [reply chapter_1/kmeans_1pluseps_approx.tex]: See how it is done in other parts

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 32]
```diff
@@ -0,0 +1,207 @@
...
+which implies $\delta \leq \frac{2}{1-2\epsilon}\norm{x-c_2}\leq 3 \norm{x-c_2}$. Using the bound on $\norm{x-c_2}$ and $\delta$, we have $\norm{x-c_1} \leq (1+ 6\epsilon)\norm{x-c_2}$ as required. 
+\end{proof}
+
+% We have already shown that for an $\epsilon$-approximate centroid set, the cost of a Voronoi partition for some pair of centroids is at most $(1+\epsilon)cost(\Pi)$ where $\Pi$ is any $2$-clustering. As the algorithm performs an exhaustive search of all pairs of centroids in the computed $\epsilon$-approximate centroid set, it is clear that it will thus try the least cost pairing and thus satisifes the $(1+\epsilon)$ approximation requirement.
+
+It remains to show that this algorithm can be performed in near-linear time. In a direct implementation of this algorithm, we must consider $|C|^2$ pairs $(c_1,c_2)$ and for each pair, we need $O(n)$ time for computing the cost of the corresponding $2$-clustering. The set $|C|$ has size $O(n \epsilon^{-d} \text{log}(1/\epsilon))$ thus implying $O(n^2)$ for a fixed $\epsilon$. However, we only need to consider an $\epsilon$-complete set of pairs of $P$ which has size $|P|=O(n \epsilon^{-d})=O(n)$.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Why does that imply O(n^2) for fixed epsilon?

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 33]
```diff
@@ -0,0 +1,207 @@
...
+
+Further, using approximate range searching, computing the approximate cost can be done in $O(\text{log }n)$ time (down from $O(n)$). As each preprocessing step takes at most $O(n \log n)$ and the main for loop takes $O(n \log n)$ for a fixed $\epsilon$, the algorithm can thus run in $O(n \log n)$.
+
+The full k-clustering algorithm builds on the intuition of the 2-clustering version and employs a recursive approach for finding the $(1+\epsilon)$ approximate.
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: apprxoimate what? There is a word missing or something at the end of this sentence

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 34]
```diff
@@ -0,0 +1,207 @@
...
+  \ENDFOR
+\ENDFOR
+\STATE For each $k^*$-tuple in $\cC^*$ with all the $C_i$ nonempty,
+we know have a single k-clustering. Now return the k-clustering with
+the smallest cost (minimum over all $k^*$-tuples and over all $k^*=2,3,...,k$.
+\end{algorithmic}
+\end{algorithm}
+
+\begin{remark}
+\textbf{The proof of running time for this method builds on the proof for approximate $2$-clustering and employs the recursive nature of the algorithm to go from $O(n \log n)$ for 2-clustering to $O(n (\log n)^k)$ for the $k$-deep recursion tree of the approximate $k$-clustering.}
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Why is this bold? Remark already italicizes it

File: chapter_1/kmeans_1pluseps_approx.tex [hunk 35]
```diff
@@ -0,0 +1,207 @@
+% Contributers: Noah Huber-Feely, Samuel Sharpe
```
Comment (all2187) [inline chapter_1/kmeans_1pluseps_approx.tex]: Overall, you did a great job in this section. However, you may want to consider breaking this section into 2 or three subsections. Maybe 1 subsection for preliminary definitions, one for the 2means result, and one for kmeans. Also try to add some intuitive explanations of the algorithms for 2means and kmeans.

**Review by all2187 — CHANGES_REQUESTED** [review]
For the sections, how about you move the two hardness of kmeans results into the relevant section (section 5). Then make a new section after the heuristics section for the approximation result (since you refer to Lloyd's method). Or make it the section before heuristics (this may make more sense) but remove references to Lloyd's method. Do whichever makes more sense to you. Also if you could add a few lines about why people use heuristics (lloyd's, kmeans++, etc) in practice even though a 1+epsilon approximation is possible, that would be great.

I've made a number of smaller comments on the PR as well. Let me know if you have any questions.

Overall though, great job!

**Comment by all2187** [comment]
I see there are some new commits, just let me know when it is ready to rereview.

**Comment by huberf** [comment]
@all2187 We have now added all the requested formatting changes and content additions. Let us know if there is anything more or we did something in a different way than you want. Thanks!

**Review by all2187 — APPROVED** [review]
Looks great! Good job!

**Pull request merged**